java/arcs/core/data/Plan.kt

```diff
@@ -21,8 +21,33 @@ import arcs.core.util.lens
 open class Plan(
     // TODO(cromwellian): add more fields as needed (e.g. RecipeName, etc for debugging)
     val particles: List<Particle>,
-    val arcId: String? = null
+    val annotations: List<Annotation>
 ) {
+    // TODO(b/155796088): get rid of these ctors and use annotations in PlanPartition and Recipe.
+    constructor(particles: List<Particle>, arcId: String? = null) :
+        this(
+            particles,
+            if (arcId != null)
+                listOf(Annotation("arcId", mapOf("id" to AnnotationParam.Str(arcId))))
+            else emptyList()
+        )
+
+    constructor(particles: List<Particle>, annotations: List<Annotation>, arcId: String) :
+        this(particles, annotations) {
+        require(this.arcId == arcId)
+    }
+
+    val arcId: String?
+        get() {
+            return annotations.find { it.name == "arcId" }?.let {
```

Comment:
An idea: you could create an `ArcIdAnnotation` class/function, to bundle up these checks and make them reusable.

It might also be helpful to make some helper methods for these things on the Annotation class, e.g. getParam could throw an error like "Expected annotation $name to have parameter $id", or something.

Reply:
that's a great idea, let me do this in a follow up pr.